cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2312,35 +2287,137 @@ class S3FileSystem::Impl : public std::enable_shared_from_this<S3FileSystem::Imp
       futures.push_back(std::move(fut));
     }
 
-    return AllComplete(futures);
+    return AllFinished(futures);
   }
 
   Status DeleteObjects(const std::string& bucket, const std::vector<std::string>& keys) {
     return DeleteObjectsAsync(bucket, keys).status();
   }
 
+  Future<> EnsureNotFileAsync(const std::string& bucket, const std::string& key) {
+    if (key.empty()) {
+      // There is no way for a bucket to be a file
+      return Future<>::MakeFinished();
+    }
+    auto self = shared_from_this();
+    return DeferNotOk(SubmitIO(io_context_, [self, bucket, key]() mutable -> Status {
+      S3Model::HeadObjectRequest req;
+      req.SetBucket(ToAwsString(bucket));
+      req.SetKey(ToAwsString(key));
+
+      ARROW_ASSIGN_OR_RAISE(auto client_lock, self->holder_->Lock());
+      auto outcome = client_lock.Move()->HeadObject(req);
+      if (outcome.IsSuccess()) {
+        const auto& result = outcome.GetResult();
+        if (result.GetContentLength() > 0 || key[key.size() - 1] != '/') {
+          return Status::IOError("Cannot delete directory contents at ", bucket, kSep,
+                                 key, " because it is a file");
+        }
```

Comment:
It seems a bit weird to have this code in a helper function named `EnsureFileAsync`.
Perhaps make this a private helper inside `DoDeleteDirContentsAsync`, or have this return a `Result<FileType>` instead?

Reply:
I changed the method to `EnsureIsDirAsync` and changed it so that it returns `Result<bool>` and moved the error message into `DeleteDirContentsAsync`.